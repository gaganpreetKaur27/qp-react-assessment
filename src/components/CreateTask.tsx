import React, { useCallback, useState } from "react";
//utils
import { CreateTaskProp } from "../utils/types/TaskTypes";
import "../utils/css/styles.css";

export const CreateTask: React.FC<CreateTaskProp> = ({ addTask }) => {
  const [value, setValue] = useState("");

  const handleAddTask = useCallback(() => {
    if (!value) return;
    addTask(value);
    setValue("");
  },[value]);

  return (
  <div className="add-task-container">
      <input
        type="text"
        className="input"
        value={value}
        placeholder="Add a new task to the list.. :)"
        onChange={(e) => setValue(e.target.value)}
      />
      <button className="add-button-style" onClick={handleAddTask}>
       Add Task
    </button>
    </div>
  );
};
