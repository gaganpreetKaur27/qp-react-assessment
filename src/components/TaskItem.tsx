import React from "react";
//utils
import { TaskProps } from "../utils/types/TaskTypes";
import "../utils/css/styles.css";

export const TaskItem: React.FC<TaskProps> = ({
  task,
  index,
  completeTask,
  removeTask,
}) => {
  return (
    <div
      className="task"
      style={{ textDecoration: task.completed ? "line-through" : "" }}
      key={index}
    >
      <span>{task.title}</span>
      <button
        className="button-container"
        style={{ background: "red" }}
        onClick={() => removeTask(index)}
      >
        x
      </button>
      <button
        data-testID="complete-btn"
        className={
          !task.completed ? "checkbox-button-pending" : "button-container"
        }
        onClick={() => completeTask(index)}
      >
        {task.completed ? "Completed" : ""}
      </button>
    </div>
  );
};
