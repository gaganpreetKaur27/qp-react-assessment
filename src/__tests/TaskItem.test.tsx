import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { TaskItem } from "../components/TaskItem";

describe("TaskItem component", () => {
  const completeTaskFn = jest.fn();
  const removeTaskFn = jest.fn();
  const mockIndex = 0;
  const taskItemData = {
    title: "",
    completed: false,
  };

  it("TaskItem matches snapshot", () => {
    const { asFragment } = render(
      <TaskItem
        task={taskItemData}
        index={0}
        completeTask={completeTaskFn}
        removeTask={removeTaskFn}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button with correct text and triggers completeTask function on click", async () => {
    const { getByTestId } = render(
      <TaskItem
        task={taskItemData}
        index={mockIndex}
        completeTask={completeTaskFn}
        removeTask={removeTaskFn}
      />
    );
    const button = getByTestId("complete-btn");
    // Act
    fireEvent.click(button);
    // Assert
    await waitFor(() => {
      // Assert
      expect(completeTaskFn).toHaveBeenCalledWith(mockIndex);
    });
  });
});
