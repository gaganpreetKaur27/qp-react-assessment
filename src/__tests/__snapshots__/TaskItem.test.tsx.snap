// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem component TaskItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="task"
  >
    <span />
    <button
      class="button-container"
      style="background: red;"
    >
      x
    </button>
    <button
      class="checkbox-button-pending"
      data-testid="complete-btn"
    />
  </div>
</DocumentFragment>
`;
